script/copy-dict.js

```diff
@@ -0,0 +1,33 @@
+// This script will copy all `.d.ts` files from `lib/` into `dist/`
+
+var fs = require('fs')
+var path = require('path')
+
+var rootDir = path.resolve(__dirname, '..')
+var distDir = path.join(rootDir, 'dist')
+var libDir = path.join(rootDir, 'lib')
+
+function* walk(dir) {
+  for (var d of fs.readdirSync(dir)) {
+    var entry = path.join(dir, d)
+    var stat = fs.lstatSync(entry)
+    if (stat.isDirectory()) yield* walk(entry)
+    else if (stat.isFile()) yield entry
+  }
+}
+
+function copy(libFilePath) {
+  var distFilePath = path.join(distDir, path.relative(libDir, libFilePath))
+  fs.mkdirSync(path.dirname(distFilePath), {recursive: true}) // Need Node.js 10.12.0+
+  fs.copyFileSync(libFilePath, distFilePath)
```

Comment:
I just found out that `mdast-util-from-markdown` requires `micromark/dist/constant/constantst`, so the types should not be compiled away. I removed the logic of `@for-script: REMOVE_ALL_THING_BELOW`.

Reply:
True, but that’s an open issue: https://github.com/syntax-tree/mdast-util-from-markdown/issues/4. I don’t think from-markdown or other folks should use it

Reply:
I know nothing about compilation, but shouldn't these imports be tree shaken by the compiler like Webpack? (Does Webpack have the ability to tree shake ES5 `require` imports?)

For example, `mdast-util-from-markdown` only uses two constants (`numericBaseDecimal` and `numericBaseHexadecimal`) from `micromark/dist/constant/constants`, so other constants except for these two constants should not appear in the final bundle. If my understanding is correct, two constants are too small to worry about.

Reply:
webpack indeed isn’t very good at shaking CJS, it can be better with [webpack-common-shake](https://github.com/indutny/webpack-common-shake).

In my experience, browserify is much better than webpack. Using browserify w/ tinyify shavings 1kb gzip off the bundle size, making micromark around 13.5kb instead of 14.5kb.

Still, no tool can do what we do in this project with [pre-evaluating](https://github.com/micromark/micromark#size--debug) code and removing all the constants, which is significant!

---

Regardless, I don’t see much value in typing these constants, and most of the utilities. I don’t consider them the API: they are not stable, they shouldn’t really be used. Instead, it makes it harder for me, as I don’t write typescript, to maintain this project (and other folks who may want to help out but don’t enjoy or know TS either).

I think you’re doing an amazing job here, but parts of of it feel a bit much to me? (See https://github.com/micromark/micromark/pull/25#issuecomment-705076345)

Reply:
> I don’t consider them the API: they are not stable, they shouldn’t really be used.  

I added `@for-script: REMOVE_ALL_THING_BELOW` back. 

> parts of of it feel a bit much to me 

Do you have more detailed instructions? Like which part of `micromark` should be considered as internal and shouldn't be exported to outside?

Reply:
I would consider `lib/*/**/*.js` private and the rest (`lib/*.js`) public. So: I see `lib/character/ascii-atext`, `lib/constant/assign`, `lib/tokenize/attention`, and the rest in subdirectories all as private.

In the case where interfaces are exposed to extensions, such as `lib/compile/html` and `lib/util/create-tokenizer`, I believe types do makes sense though.

Reply:
Just to be clear: all the files below are public, and the rest are all private. Am I getting this right?

```
lib/constructs.js
lib/index.js
lib/parse.js
lib/postprocess.js
lib/preprocess.js
lib/stream.js
lib/compile/html.js
lib/util/create-tokenizer.js 
```

------

Reply:
I’d say these are public:

```
lib/index.js
lib/parse.js
lib/postprocess.js
lib/preprocess.js
lib/stream.js
```

^-- this is what folks will deal with

Then there are things for folks making extensions. They will define SyntaxExtensions and HtmlExtensions, where they define constructs with initializers, resolvers, states, and the like, which can access stuff coming from compile/html and create-tokenizer

Reply:
However I some plugins do import more stuff from `micromark`:


https://github.com/syntax-tree/mdast-util-from-markdown/blob/a9d53edd090b71a4d50d4fababaad63d44cadaa2/index.js#L11-L12


Are these imports valid? Or these imports should be removed?

Reply:
Yes, that’s correct! But I don’t think others should do the same thing. And I think those should be removed in the future.

mdast-util-from-markdown is highly coupled to micromark. It is more a fork, or even the same thing, using hidden internals. It’s not a plugin/extension.

Some of the extensions I wrote also [use utilities](https://github.com/micromark/micromark-extension-footnote/blob/a9bbcb93fda89765851e5819ec8c8115a2108069/index.js#L3-L9) from micromark core. I don’t think that’s good though or that others should do it.

Reply:
Removed all `.d.ts` for private files. Now the whole repository only has the following `.d.ts` files:

- public API
  - ./lib/index.d.ts
  - ./lib/parse.d.ts
  - ./lib/postprocess.d.ts
  - ./lib/preprocess.d.ts
  - ./lib/steam.d.ts
- Some utility type to making extensions like `SyntaxExtensions`, `HtmlExtensions`, `Construct`, `Tokenizer`, etc
  - ./lib/shared-types.d.ts
- Some constant definitions:
  - ./lib/character/codes.d.ts
  - ./lib/character/values.d.ts
  - ./lib/constant/constants.d.ts
  - ./lib/constant/types.d.ts

Reply:
thanks for understanding, @ocavue!

To add: I think it’s wise to start small, and go from there! Maybe some of these things in utilities can be made into external tools or a public API later!